lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+            restage (bool): ``True`` if forcing Spack to restage the package
+                source, otherwise ``False``
+        """
+        # Make sure the package is ready to be locally installed.
+        self._ensure_install_ready(task.pkg)
+
+        # Skip file system operations if we've already gone through them for
+        # this spec.
+        if task.pkg.unique_id in self.installed:
+            # Already determined the spec has been installed
+            return
+
+        # Determine if the spec is flagged as installed in the database
+        try:
+            rec = spack.store.db.get_record(task.pkg.spec)
+            installed_in_db = rec.installed if rec else False
+        except KeyError:
+            rec = None
+            installed_in_db = False
+
+        # Make sure the installation directory is in the desired state
+        # for uninstalled specs.
+        partial = False
+        if not installed_in_db and os.path.isdir(task.pkg.spec.prefix):
+            if not keep_prefix:
+                task.pkg.remove_prefix()
+            else:
+                tty.debug('{0} is partially installed'
+                          .format(task.pkg.unique_id))
+                partial = True
+
+        # Destroy the stage for a locally installed, non-DIYStage, package
+        if restage and task.pkg.stage.managed_by_spack:
+            task.pkg.stage.destroy()
+
+        if not partial and self.layout.check_installed(task.pkg.spec):
+            self._update_installed(task)
+
+            # Only update the explicit entry once for the explicit package
+            if task.pkg.unique_id == self.pkg.unique_id:
+                _update_explicit_entry_in_db(task.pkg, rec, True)
+
+            # In case the stage directory has already been created, this
+            # check ensures it is removed after we checked that the spec is
+            # installed.
+            if not keep_stage:
+                task.pkg.stage.destroy()
+
+    def _check_last_phase(self, **kwargs):
+        """
+        Ensures the package being installed has a valid last phase before
+        proceeding with the installation.
+
+        The ``stop_at`` argument is removed from the installation arguments.
+
+        Args:
+            kwargs:
+              ``stop_at``': last installation phase to be executed (or None)
+        """
+        self.pkg.last_phase = kwargs.pop('stop_at', None)
+        if self.pkg.last_phase is not None and \
+                self.pkg.last_phase not in self.pkg.phases:
+            tty.die('\'{0}\' is not an allowed phase for package {1}'
+                    .format(self.pkg.last_phase, self.pkg.name))
+
+    def _cleanup_all_tasks(self):
+        """Cleanup all build tasks to include releasing their locks."""
+        for pkg_id in self.locks:
+            self._release_lock(pkg_id)
+
+        for pkg_id in self.failed:
+            self._cleanup_failed(pkg_id)
+
+        ids = list(self.build_tasks.keys())
```

Comment:
I think this will work even without using `keys` explicitly:
```suggestion
        ids = list(self.build_tasks)
```

Reply:
Removed